internal/engine/ingester/artifact/artifact.go

```diff
@@ -336,21 +341,30 @@ func branchFromRef(ref string) string {
 	return ""
 }
 
-func signerIdentityFromCertificate(c *certificate.Summary) string {
-	if c.BuildSignerURI != "" {
-		// TODO(puerco): We should not be trimmin the tag from the signer identity
-		// I'm leavin this part for now as we are capturing the tag and branch
-		// elsewhere but we should improve this.
-		//
+// signerIdentityFromCertificate returns the signer identity. When the identity
+// is a URI (from the BuildSignerURI extension or the cert SAN), we return only
+// the URI path component. We split it this way to ensure we can make rules
+// more generalizable (applicable to the same path regardless of the repo for example).
+func signerIdentityFromCertificate(c *certificate.Summary) (string, error) {
+	var builderURL string
+	switch {
+	case c.BuildSignerURI != "":
 		// Find the index of '@' to isolate the file name
-		b, _, _ := strings.Cut(c.BuildSignerURI, "@")
-		return b
+		builderURL = c.BuildSignerURI
+	case c.SubjectAlternativeName.Value != "" && c.SubjectAlternativeName.Type == certificate.SubjectAlternativeNameTypeURI:
+		builderURL = c.SubjectAlternativeName.Value
+	default:
+		// Return the SAN in the cert as a last resort. This handles the case when
+		// we don't have a signer identity but also when the SAN is an email
+		// when a user authenticated using an OIDC provider.
+		// Any other SAN types are returned verbatim
+		return c.SubjectAlternativeName.Value, nil
+	}
 
-	} else if c.SubjectAlternativeName.Value != "" {
-		// This is the use case where there is no build signer URI but there is a subject alternative name
-		// Usually this is the case when signing through an OIDC provider. The value is the signer's email identity.
-		return c.SubjectAlternativeName.Value
+	uString, _, _ := strings.Cut(builderURL, "@")
```

Comment:
@puerco this is much better than what we had and I like that the code is generic, but in this case I wonder if making the code a little GH specific would go a long way towards better usability, because the returned path still contains the org and repo paths (`jakubtestorg/demo-repo-go/.github/workflows/build-image-signed-cosign-static.yml` in my case).

Do you think it would be too much of a hack to check if the extensions.Issuer is `https://token.actions.githubusercontent.com` and in that case strip also `GithubWorkflowRepository`? Is it possible to forge signatures with that data that we would successfully verify?

Reply:
If we're validating the signature and it comes from a public (or GH) sigstore, it should be fine to make that assumption. That is, to check the issuer and get the relevant metadata from there. Public sigstore has a subset of IdPs they support and they do check the issuer before giving out a Fulcio certificate. This would be a different case if we'd be verifying with a private sigstore though.